### PR TITLE
overview of data persistence

### LINKED ISSUE #4750: overview of persisting data
We don't have a doc that covers workspaces, caching and artifacts and when to use each, as per this blog post. Might be a nice overview to have somewhere as this info is quite spread out in docs right now:

https://circleci.com/blog/persisting-data-in-workflows-when-to-use-caching-artifacts-and-workspaces/

### PR DESCRIPTION
**Pull request by rosieyohannan**
Fixes #4750 and #4749

### CODE REVIEW

**Review by melyalex — APPROVED** [review]

**Comment by rosieyohannan** [comment]
I'm very open to changing the title to make it easier to find 🤷

File: jekyll/_cci2/persist-data.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+This guide gives an overview of the various ways to persist data within and beyond your CircleCI builds. There are a number of ways to move data into, out of and between jobs, and persist data for future use. Using the right feature for the right task will help speed up your builds and improve repeatability and efficiency.
+
+* TOC
+{:toc}
+
+## Caching Strategies
+
+![caching data flow]({{ site.baseurl }}/assets/img/docs/caching-dependencies-overview.png)
+
+Caching persists data between the same job in different Workflow builds, allowing you to reuse the data from expensive fetch operations from previous jobs. After an initial job run, future instances will run faster as they will not need to redo the work. A prime example is package dependency managers such as Yarn, Bundler, or Pip. With dependencies restored from a cache, commands like yarn install will only need to download new dependencies, if any, and not redownload everything on every build.
```
Comment (teesloane) [inline jekyll/_cci2/persist-data.md:19]:
```suggestion
Caching persists data between the same job in different Workflow builds, allowing you to reuse the data from expensive fetch operations from previous jobs. After an initial job run, future instances will run faster as they will not need to redo the work (provided your cache has not been invalidated). A prime example is package dependency managers such as Yarn, Bundler, or Pip. With dependencies restored from a cache, commands like yarn install will only need to download new dependencies, if any, and not redownload everything on every build.
```

File: jekyll/_cci2/persist-data.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+## Using Workspaces
+
+![workspaces data flow]( {{ site.baseurl }}/assets/img/docs/workspaces.png)
+
+When a workspace is declared in a job, files and directories can be added to it. Each addition creates a new layer in the workspace filesystem. Downstream jobs can then use this workspace for their own needs or add more layers on top.
+
+Unlike caching, workspaces are not shared between runs as they no longer exists once a workflow is complete. The one exception is when workflows are rerun.
+
+**Workspaces are stored for up to 15 days.**
```
Comment (teesloane) [inline jekyll/_cci2/persist-data.md:35]: The line prior to this one states that a workspace no longer exists once a workflow is complete. If they are stored for up to 15 days... does that mean, only in the context of "rerunning a workflow" is that data made available?

File: jekyll/_cci2/persist-data.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+For more information on using workspaces to persist data throughout a workflow, see the [Workflows]({{site.baseurl}}/2.0/workflows/#using-workspaces-to-share-data-among-jobs) guide. Also see the [Deep Diving into CircleCI Workspaces](https://circleci.com/blog/deep-diving-into-circleci-workspaces/) blog post.
+
+## Using Artifacts
+
+![artifacts data flow]( {{ site.baseurl }}/assets/img/docs/Diagram-v3-Artifact.png)
+
+Artifacts are used for longer-term storage of the outputs of your pipelines. For example if you have a Java project, your build will most likely produce a `.jar` file of your code. This code will be validated by your tests. If the whole build/test process passes, then the output of the process (the `.jar`) can be stored as an artifact. The `.jar` file is available to download from our artifacts system long after the workflow that created it has finished.
+
+If your project needs to be packaged, say an Android app where the `.apk` file is uploaded to Google Play, that’s a great example of an artifact. Many users take their artifacts and upload them to a company-wide storage location such as Amazon S3 or Artifactory.
```
Comment (teesloane) [inline jekyll/_cci2/persist-data.md:45]:
```suggestion
If your project needs to be packaged, say an Android app where the `.apk` file is uploaded to Google Play, you would likely wish to store it as an artifact. Many users take their artifacts and upload them to a company-wide storage location such as Amazon S3 or Artifactory.
```
Just some opinionated wording changes.

**Review by teesloane — CHANGES_REQUESTED** [review]
Nice work! I could see this being a "top level" document that branches into these features; possibly something we might want to consider for a sidebar revamp.

File: jekyll/_cci2/persist-data.md (hunk 2 above)
Comment (rosieyohannan) [reply jekyll/_cci2/persist-data.md:35]: Yeah good question - I think that's correct but will check with eng

File: jekyll/_cci2/persist-data.md (hunk 3 above)
Comment (rosieyohannan) [reply jekyll/_cci2/persist-data.md:45]: Thanks so much for the review @teesloane!

**Review by thomasr8 — APPROVED** [review]
Looks great @rosieyohannan 👍

File: jekyll/_cci2/persist-data.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+For more information see the [Caching Dependencies]({{site.baseurl}}/2.0/caching/) guide.
+
+## Using Workspaces
+
+![workspaces data flow]( {{ site.baseurl }}/assets/img/docs/workspaces.png)
+
+When a workspace is declared in a job, files and directories can be added to it. Each addition creates a new layer in the workspace filesystem. Downstream jobs can then use this workspace for their own needs or add more layers on top.
+
+Unlike caching, workspaces are not shared between runs as they no longer exists once a workflow is complete. The one exception is when workflows are rerun.
```
Comment (rosieyohannan) [inline jekyll/_cci2/persist-data.md:33]:
```suggestion
Unlike caching, workspaces are not shared between pipeline runs. The only time a workspace can be accessed after the pipeline has run is when a workflow is rerun within the 15 day limit.
```

File: jekyll/_cci2/persist-data.md (hunk 2 above)
Comment (rosieyohannan) [reply jekyll/_cci2/persist-data.md:35]: https://circleci.slack.com/archives/CNPG1R4QZ/p1601650871053200

**Review by melyalex — APPROVED** [review]

**Review by melyalex — APPROVED** [review]

**Review by melyalex — APPROVED** [review]

**Review by melyalex — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Review by abretmounet — APPROVED** [review]

**Comment by rosieyohannan** [comment]
reopening these changes in #4808

**Pull request closed without merging**